File: project_euler/Problem049/Problem049.c
#include <stdio.h>
#include <math.h>

static int is_prime(int value);
static int check_digit(int value, int *nexts, int *digits);

void Problem049(void) {
    int target = 3330;
    int other = 1487;
    int terms[2];
    int digits[10];
    for (int i = 10000 - (target << 1) - 1; i > 1000; i -= 2) {
        if (is_prime(i) && i != other) {
            terms[0] = i + target;
            terms[1] = terms[0] + target;
            if (is_prime(terms[0]) && is_prime(terms[1]) && check_digit(i, terms, digits)) {
                printf("%d%d%d\n", i, terms[0], terms[1]);
                return;
            }
        }
    }
}

static int is_prime(int value) {
    if (!(value & 1)) {
        return (2 == value) ? 1 : 0;
    }

    int border = (int) sqrt(value);
    for (int i = 3; i <= border; i += 2) {
        if (!(value % i)) {
            return 0;
        }
    }
    return (1 != value) ? 1 : 0;
}

static int check_digit(int value, int *terms, int *digits) {
    for (int i = 0; i < 10; ++i) {
        digits[i] = 0;
    }
    while (value) {
        digits[value % 10] = 1;
        value /= 10;
    }

    for (int i = 0; i < 2; ++i) {
        int term = terms[i];
        while (term) {
            if (!digits[term % 10]) {
                return 0;
            }
            term /= 10;
        }
    }
    return 1;
}

File: project_euler/Problem049/Problem049.java
package project_euler;

import java.util.Arrays;

public class Problem049 {
    public void solution() {
        int target = 3330;
        int other = 1487;
        int[] terms = new int[2];
        boolean[] digits = new boolean[10];
        for (int i = 10000 - (target << 1) - 1; i > 1000; i -= 2) {
            if (isPrime(i) && i != other) {
                terms[0] = i + target;
                terms[1] = terms[0] + target;
                if (isPrime(terms[0]) && isPrime(terms[1]) && checkDigit(i, terms, digits)) {
                    System.out.printf("%d%d%d\n", i, terms[0], terms[1]);
                    return;
                }
            }
        }
    }

    private boolean isPrime(int value) {
        if (1 != value % 6 && 5 != value % 6) {
            return 2 == value || 3 == value;
        }

        int border = (int) Math.sqrt(value);
        for (int i = 5; i <= border; i += 6) {
            if (0 == value % i || 0 == value % (i + 2)) {
                return false;
            }
        }
        return 1 != value;
    }

    private boolean checkDigit(int value, int[] terms, boolean[] digits) {
        Arrays.fill(digits, false);
        while (0 != value) {
            digits[value % 10] = true;
            value /= 10;
        }

        for (int term : terms) {
            while (0 != term) {
                if (!digits[term % 10]) {
                    return false;
                }
                term /= 10;
            }
        }
        return true;
    }
}
